### PR TITLE
core/cli: fix empty `.buildinfo` files with `--incremental`

### PR DESCRIPTION
**Pull request by chriskrycho**
@dfreeman encountered a scenario where the `.buildinfo` file was generated but empty. The first commit here is a first pass to verify that it does get generated at least sometimes. Unfortunately, it does not reproduce the bug; but it *does* tell us that this at least *generally* works.

### CODE REVIEW

**Comment by dfreeman** [comment]
Closing the loop on our conversation: the file wasn't literally empty, but it only contained a version marker without any actual build info:

```json
{"version":"4.8.2"}
```

I'll poke at it some next week if I get the opportunity

**Comment by chriskrycho** [comment]
Ah, yes. We need a slightly more robust test than this; I originally used a snapshot and that had the full contents, so there is something else different between the scenario our tests hit and the real-world failure!

**Comment by chriskrycho** [comment]
Update here: I identified that the problem is we were *testing* `--build` but what was broken was a normal 'check' pass. The new tests exercise it correctly. Unfortunately, I haven't been able to figure out yet what we *do* need to do to actually make it work. I’m pushing a WIP commit that shows one path which *doesn’t* work, and because it doesn’t I am mildly stumped.

**Comment by chriskrycho** [comment]
My notes from the commit message in the latest:

> I would expect this to work: it simply wires up the compiler the *exact* same way we do for `--build`, but with `--incremental`, but... no. It's possible (perhaps even likely) that we need to do some extra wiring; knowing for sure will likely require reading the TS source code. As it does for these things.

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "release-it": "echo \"Running release-it via yarn breaks publishing! Use npx or a Volta global installation.\""
   },
   "volta": {
-    "node": "14.19.1",
+    "node": "16.17.1",
```
Comment (chriskrycho) [inline package.json:18]: Ignore this; I'll pull it out of this and check it in on a separate PR and just make sure CI still tests—I spent another 15 minutes yesterday trying to figure out whyyyyyy I couldn't get anything to work with Vitest before remembering that it was the "bump to Node 16 so ESBuild works natively" dance.

**Pull request merged**